### DISCUSSION TITLE
История прослушивания без пропусков и скипов [устарело]

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#53] [Избранное]
## Этот метод перестал работать с новыми версиями приложений Spotify, переходите на [новый метод](https://github.com/Chimildic/goofy/discussions/230).
>[!Warning]
Разработчики Spotify изменили локальный файл в котором хранилась история прослушивания, но если очень хочется 
пользоваться данным методом то нужно понижать версию вашего приложения, для Windows это <kbd>1.1.84</kbd> и ниже, что 
бы сделать понижение версии и заблокировать обновления воспользуйтесь [этим](https://github.com/amd64fox/Rollback-Spotify). Для андроида попробуйте эту версию [8.7.20.1261 Build: 85467771](https://4pda.to/forum/index.php?showtopic=248440&st=9900#entry67736469) или ниже.

Всем привет, выкладываю альтернативный вариант для ведения истории прослушивания, спасибо @Chimildic, весь код для этого модуля был написан им.  Сделано было это еще перед новым годом, как-то в суматохе не подумал сюда сразу выставить, исправляюсь.

### **Краткое описание.**

Ведение истории прослушивание треков на **Android** и **Windows**.

**Плюсы:** 
- треки добавляются с 0 сек.
- в историю будет добавлен именно тот трек который был прослушан, так как у нас есть его точный id.

**Минусы:**
- треки идут без даты.
- нужно настраивать выгрузку в google диск через то устройство на котором вы слушаете.  

### **Суть схемы работы.**
 Spotify в клиентах для windows, android в открытом доступе, локально хранит историю прослушивания в обычном тексте, этот файл и нужно выслать goofy для его последующей обработки в историю.
 

### **Инструкция первого запуска.**
 **1.** Копируем код ниже в отдельный файл, например `my-module.gs`, что бы жил независимо от обновлений основной библиотеки.

```javascript
const HistoryManager = (function () {
    const FILES = ['SpotifyHistoryPC', 'SpotifyHistoryPhone'];
    const rootFolder = DriveApp.getFoldersByName(User.id).next();
    return {
        getTracks: getTracks,
        removeTracks: removeTracks,
        readTrackIds: readTrackIds,
    };

    function getTracks() {
        return SpotifyRequest.getFullObjByIds('tracks', readTrackIds(), 50);
    }

    function removeTracks(tracks) {
        let trackHistoryIds = readTrackIds();
        return tracks.filter((track) => !trackHistoryIds.includes(track.id));
    }

    function readTrackIds() {
        let tracks = [];
        for (let i = 0; i < FILES.length; i++) {
            let data = tryParseJSON(getFile(FILES[i]));
            if (data.play_history && data.play_history.tracks) {
                let validTracks = data.play_history.tracks.filter(t => t.uri.includes('spotify:track'));
                Combiner.push(tracks, validTracks);
            }
        }
        let trackIds = tracks.map((track) => track.uri.replace('spotify:track:', ''));
        return Array.from(new Set(trackIds));
    }

    function tryParseJSON(file) {
        if (!file) return {};
        try {
            let dataAsString = tryGetBlob(file);
            let hashIndex = dataAsString.indexOf('#');
            return JSON.parse(dataAsString.substring(hashIndex + 1));
        } catch (error) {
            Admin.printError(error.stack);
            return {};
        }
    }

    function tryGetBlob(file) {
        if (!file) return '';
        try {
            return file.getBlob().getDataAsString();
        } catch (error) {
            Admin.printError('При получении данных из файла произошла ошибка\n', error.stack);
            Admin.pause(5);
            return tryGetBlob(file);
        }
    }

    function getFile(filename) {
        let files = rootFolder.getFilesByName(filename);
        if (files.hasNext()) {
            return files.next();
        }
    }
})();
```

**2.** Далее например используем вариант для android, для передачи файла с телефона я использую tasker.

- открываем tasker.
- создаем новую задачу. (`TASKS`)
- создаем новое действие.
- выбираешь из списка `Google Drive`. (у меня он самый первый)
- потом `GD Upload`.
- далее заполняем строки в нашем действие `GD Upload`. 
  - `Google drive Account` - ваш аккаунт google.
  -  `Data / File` - путь к файлу по умолчанию `Android/data/com.spotify.music/files/spotifycache/Users/ваш-ID-user/context_player_state_restore` (Если вдруг там пустая папка, то дайте необходимые разрешения для таскера или проверьте туже директорию но на sd-карте)
  - `Remote File Name`-  имя файла после его выгрузки в google, по умолчанию `SpotifyHistoryPhone`. 
  - `Remote Folder` - в какую папку выгружать файл на гугл диске, по умолчанию `Goofy data/id-аккаунта` .
  - поставить галку на `Overwrite if Exists`. (каждый раз перезаписывать файл)
- Задача готова, остаётся выбрать себе триггер по душе и связать их. Я, например сделал себе триггер `Notification Removed Spotify` то есть файл с историей будет выгружаться каждый раз после того как любое уведомление в spotify будет закрыто, а это всегда будет наш плеер по сути.
  - Идем в `PROFILES` 
  - Выбираем `Событие`
  -  `Интерфейс` => `Notification Removed`
  - В `Приложение-владелец` выбираем Spotify, все сохраняем, таскер настроен.
  
**3.** Создаем функцию для Goofy, идем в `main` и создаем новую функцию `test`
```javascript
function test() {
    let tracks = HistoryManager.getTracks();

 Playlist.saveAsNew({
      name: 'test-history',
      tracks: tracks,
      });
```
**4.** Проверяем работу, идем в Spotify слушаем пару или больше треков, закрываем Spotify, что бы сработал триггер или другое действие которое назначено у вас. 
**5.** Если выгрузка сработала правильно в Google диске, появится файл `SpotifyHistoryPhone`.
**6.** Идем в `Goofy` и запускаем нашу функцию `test`, проверяем наш новый плейлист test-history, обычно если не фильтровать массив то выходит от 200 до 500 треков истории.
**7.** Последний пункт обвязка функции, в начальном массиве истории дубликаты уже удалены по id, так что фильтровать на повторы в сам массиве новой истории не нужно, я организовал сбор хранение истории через Cache.

```javascript
function test() {
    let tracks_new = HistoryManager.getTracks(); // Беру свою новую историю.
    let tracks_old = Cache.read('History.json'); // Беру свою основную историю.
    Filter.removeTracksid(tracks_new, tracks_old);  // Удаляю дубликаты из новой истории по id.
    Cache.compressTracks(tracks_new); // Cache.compress по желанию.
    Cache.append('History.json', tracks_new);  // Все что осталось добавляю в основную историю.
    console.log('Количество добавленных треков в историю:', tracks_new.length);
```

Так же для моей истории был создан `Filter.removeTracksid`  тот же `Filter.removeTracks` как и основной но фильтрует только по id, кому это не надо могут использовать обычный вариант из основной библиотеки, кто понял зачем тут фильтр по id то вот его код:
```javascript
Filter.removeTracksid = function (sourceArray, removedArray, invert = false) {
        let removedIds = removedArray.map((item) => item.id);
        let filteredTracks = sourceArray.filter((item) => {
            return invert ^ !removedIds.includes(item.id);
        });
        Combiner.replace(sourceArray, filteredTracks);
    };
```
Так же как и основной код можно добавить в `my-module.gs`.
Дальше уже обвязку объяснять не буду, чисто субъективщина пойдет, думаю справитесь.

В заключении добавлю, что способ для Пк версии не особо отличается, просто найдите вариант так же отправлять файл в goofy, для себя выбрал программу `Air Explorer Pro` которую настроил параллельно с планировщиком windows. Триггер для самой функции так же настраивайте под себя, я выбрал на каждый час. 
 Пользуюсь данным способом уже 2 месяца, работает все как часы. Всем спасибо за внимание, и еще раз выражаю огромную благодарность разработчику проекта @Chimildic

**Comment by amd64fox** [comment]
Появилась ошибка после того как я зашел в этот [плейлист ](https://open.spotify.com/playlist/67vSsgklrRm6UJrAxrfl1b?si=RWPgnZeIRBiQXZB1VzwfoQ), эти треки как оказалось загружены самим пользователем и к медиатеки spotify не имеют отношения, прослушать их я не мог, но как то они залетали в файл context_player_state_restore и создают ошибку. Находятся они выше списка самой истории.


```
28 февр. 2021 г., 14:04:01	Ошибка	URL: https://api.spotify.com/v1/tracks/?ids=60Acte337xl55xgrAssHHi,1blwkKwkkSsCzwRobzqchW,3ck3j7LBLZA3TQ3WD7IccV,5i3Y6ORNtuOISoZw3axeuT,14wFoE4F2zckFz6IX8IgxB,2HCNq6LmzSNNZPsOBvoaSP,0ORj6JI4rY1OJFvvDUOz6N,0Idfl55oXvdwXm0J8MApHd,7MAnkoPF0Rf0GoXQe4h7BR,4rG6FapPRcgh7qFOswAUBz,3tNS6VXDvDhhuZV91yMDzx,2S60az586WN3qXpkzSFS4s,708VXsxOatUkOgph1v2roD,3zkgcBeYReqWqgsrXSXQho,6NGxy0MVisIbDTUWEGmq6b,5obNsqJlTL8AlSN1Ira0JO,7iOAFr8h2bggr7iXn0LE1c,3FdfMNKHU83gjiPF1DpBqw,3ayxE8wkINUsw3KO6lQjkm,7F67wuFOvqsj7eKCtYlihB,3Q61YzBn4025q6HizYgBLV,4FNikIH3gKUeuKqmeWaYZ7,3Ic5M9jvauvUh6Nw7vJWTL,2ooxeo6DjgYg27xZKxeYjv,6m4NAYdWwnkhefxXn7p6Yl,5Y9Zq3IMjFN3itCwoNJTik,spotify:local:Mark+Morgan:Vault+Archives:Khans+Of+The+New+California:202,0y8JDa79oDxilo8oq3ws2z,2EbiuwCYjV6V6OJ1u24siF,3SVZ4PjQKLFAgIvAQnULRS,0dA2HIxFnBXsZtsJO7jF6m,0Ib295LFDrAsOhGHzq1cUF,6imWTAoXsDuWypXlHJtErI,6LBXgBDohHyvUtIajQEJEo,0EF2mNCpYwYZIdgJxAU31D,35NyjYJFaJxqbUF2y0jWl1,2Bc4llhjJBW77I552RgA3L,4HBo946zIZAYhyzlZ7hfGo,1sqlv6QcCxB2JL3xmyn84G,0W3Fv89IF7mdPBY8glohzu,0jdUJ9kxGqTfAlRpg427ki,0wtYjqlEz4YRKo4oqcMTYv,7rawVLqPE6iLx79xUmmEkE,2F6tPnJC6m8SRW7LBiVtc6,7AcKOoTs1J2onMegQwKkUS,2wp62NkJotkrjRkbMsNsX6,2T6XjFBpY0eD8XC92VEpzV,2QBAiYg8cPRYOzLhGVgYOZ,2GMVFDUyBScHFjYlajlF0A,6jtDxzB1Uc6cZVQSOL78Ij 
Code: 400 
Content: {
  "error" : {
    "status" : 400,
    "message" : "invalid id"

```

**Reply by Chimildic** [reply]
Ошибка прерывает выполнение последующего кода?

**Reply by amd64fox** [reply]
да

**Reply by amd64fox** [reply]
То есть не так, скрипт выполняется с этой ошибкой, но не выводит новые прослушанные треки.

**Reply by amd64fox** [reply]
ххмм странно, попробовал раз 10 запускать вручную,  ошибку показывает, треки начал выводит через раз....

**Reply by Chimildic** [reply]
Запустил, `getTracks` не смотря на ошибку не прерывает скрипт и позволяет работать дальше.

**Reply by amd64fox** [reply]
А зачем он берет треки до фразы `play_history` ? Вся же история идет после

**Reply by Chimildic** [reply]
Треки только отсюда `data.play_history.tracks`

**Reply by amd64fox** [reply]
да, треки дергает те которые надо, я имею в виду он что-то делает с треками которые выше

**Reply by Chimildic** [reply]
Нет, не делает

**Reply by amd64fox** [reply]
В общем потестил эту ошибку более подробно и все таки она не дает выводить новые треки, при условии, что был прослушан локальный трек, если не слушать их вообще то проблем нет, но если такой трек был прослушан то он добавляется в файл именно в ту часть где хранится история, начинается такой локальный трек например вот так: `{"uid":"889","uri":"spotify:local:Nena:99+Luftballons:99+Luftballons:233","metadata":{}}`
Возможно сделать исключение если  строка трека содержит spotify:local: ?

**Reply by Chimildic** [reply]
Обновил функцию `readTrackIds` в первом посте, попробуйте с ней

**Reply by amd64fox** [reply]
Огромнейшее спасибо, все работает.

**Comment by amd64fox** [comment]
Начиная с версии 1.6.2 я так понял поменялся путь по умолчанию, теперь он вида "Goofy data/id-аккаунта" вместо "Goofy data"

В итоге история встала. У себя я путь поправил руками, что бы история дальше крутилась, а в целом наверное можно подставить какую-то переменную с уникальным id вместо старой папки "goofy data".

**Reply by Chimildic** [reply]
Попробуйте так
```js
const HistoryManager = (function () {
    const FILES = ['SpotifyHistoryPC', 'SpotifyHistoryPhone'];
    const rootFolder = DriveApp.getFoldersByName(User.id).next();
    return {
        getTracks: getTracks,
        removeTracks: removeTracks,
        readTrackIds: readTrackIds,
    };

    function getTracks(){
        return SpotifyRequest.getFullObjByIds('tracks', readTrackIds(), 50);
    }

    function removeTracks(tracks){
        let trackHistoryIds = readTrackIds();
        return tracks.filter((track) => !trackHistoryIds.includes(track.id));
    }

    function readTrackIds() {
        let tracks = [];
        for (let i = 0; i < FILES.length; i++){
            let data = read(FILES[i]);
            if (data.play_history && data.play_history.tracks){
                let validTracks = data.play_history.tracks.filter(t => t.uri.includes('spotify:track'));
                Combiner.push(tracks, validTracks);
            }
        }
        return dedup(tracks.map((track) => {
            return track.uri.replace('spotify:track:', '');
        }));
    }

    function dedup(trackIds){
        return Array.from(new Set(trackIds));
    }

    function read(filename) {
        return tryParseJSON(getFile(filename));
    }

    function tryParseJSON(file) {
        if (!file) return {};
        try {
            let dataAsString = file.getBlob().getDataAsString();
            let hashIndex = dataAsString.indexOf('#');
            return JSON.parse(dataAsString.substring(hashIndex + 1));
        } catch (e) {
            console.error(e, e.stack, file.getBlob().getDataAsString());
            return {};
        }
    }

    function getFile(filename) {
        let files = rootFolder.getFilesByName(filename);
        if (files.hasNext()) {
            return files.next();
        }
    }
})();
```

**Reply by Chimildic** [reply]
Диск не сыпет ошибками при чтении/записи? Тут не совсем актуальный способ обращения с ним

**Reply by amd64fox** [reply]
Работает, спасибо.
Да, ошибки с диском бывают, в среднем это где-то 2-3 в неделю.

**Comment by Chimildic** [comment]
Хотел поменять чтение файла на функцию из `Cache`, но `#` в начале файла истории портит всю картину. Из-за чего приходится дублировать несколько внутренних функций. Встраивать проверку символа хэша в основной код выглядит как то странно. 
Такой код отработает? 
```js
const HistoryManager = (function () {
    const FILES = ['SpotifyHistoryPC', 'SpotifyHistoryPhone'];
    const rootFolder = DriveApp.getFoldersByName(User.id).next();
    return {
        getTracks: getTracks,
        removeTracks: removeTracks,
        readTrackIds: readTrackIds,
    };

    function getTracks() {
        return SpotifyRequest.getFullObjByIds('tracks', readTrackIds(), 50);
    }

    function removeTracks(tracks) {
        let trackHistoryIds = readTrackIds();
        return tracks.filter((track) => !trackHistoryIds.includes(track.id));
    }

    function readTrackIds() {
        let tracks = [];
        for (let i = 0; i < FILES.length; i++) {
            let data = tryParseJSON(getFile(FILES[i]));
            if (data.play_history && data.play_history.tracks) {
                let validTracks = data.play_history.tracks.filter(t => t.uri.includes('spotify:track'));
                Combiner.push(tracks, validTracks);
            }
        }
        let trackIds = tracks.map((track) => track.uri.replace('spotify:track:', ''));
        return Array.from(new Set(trackIds));
    }

    function tryParseJSON(file) {
        if (!file) return {};
        try {
            let dataAsString = tryGetBlob(file);
            let hashIndex = dataAsString.indexOf('#');
            return JSON.parse(dataAsString.substring(hashIndex + 1));
        } catch (error) {
            Admin.printError(error.stack);
            return {};
        }
    }

    function tryGetBlob(file) {
        if (!file) return '';
        try {
            return file.getBlob().getDataAsString();
        } catch (error) {
            Admin.printError('При получении данных из файла произошла ошибка\n', error.stack);
            Admin.pause(5);
            return tryGetBlob(file);
        }
    }

    function getFile(filename) {
        let files = rootFolder.getFilesByName(filename);
        if (files.hasNext()) {
            return files.next();
        }
    }
})();
```

**Reply by amd64fox** [reply]
UPD. Скопипастил криво, работает.

А нельзя обрезать файл истории до фразы `play_history` или до другой условной границы и уже работать с тем что после ?

**Reply by Chimildic** [reply]
Проблема в другой области. `Cache` не сможет прочитать такой файл. И любая адаптация файла бессмысленна. Все равно какие-то действия с файлом вне `Cache` нужны. Лучше уж сразу все в одном месте. А вписывать в основной код странно. Слишком уж редкий случай.

**Reply by amd64fox** [reply]
> Слишком уж редкий случай.

Эт точно 😆

**Reply by amd64fox** [reply]
> Диск не сыпет ошибками при чтении/записи? 

Почти месяц прошел, ошибки пропали, и тут и в основной библиотеке тоже, благодарю.

**Comment by amd64fox** [comment]
Начиная с `1.1.85` десктопной версии, файл `context_player_state_restore` изменился для некоторых аккаунтов, в новом варианте нет полной истории и где она сейчас пока сложно сказать, беглый просмотр не дал результатов.

**Reply by amd64fox** [reply]
Что бы вернуть привычный файлик `context_player_state_restore` нужно перейти в `AppData\Roaming\Spotify`, открыть файл `Spotify.exe` например с помощью notepad++, в нем найти текст `enable_protobuf_state_restore`, заменить например последнюю букву `e` на `a` и сохранить, перезайти в клиент, прослушать несколько треков, выключить клиент, после этого у меня вернулся привычный файл.

**Comment by Bondian** [comment]
До недавнего времени всё работало прекрасно. Но в последние дни функция с историей стала сбоить:

```
16:54:28	Примечание	Выполнение начато
16:54:32	Ошибка	SyntaxError: Unexpected token  in JSON at position 0
    at JSON.parse (<anonymous>)
    at tryParseJSON ((   2 my-module (historyManager):37:25)
    at readTrackIds ((   2 my-module (historyManager):22:24)
    at Object.getTracks ((   2 my-module (historyManager):11:57)
    at HISTORY (1 ( ИСТОРИЯ ) 15 минут:4:35)
    at __GS_INTERNAL_top_function_call__.gs:1:8 

Описание и решение ошибок: https://chimildic.github.io/goofy/#/errors
16:54:37	Информация	Число запросов 2
16:54:35	Примечание	Выполнение `завершено
```
Я так понимаю - выше описанная проблема с файлом "context_player_state_restore" меня коснуться не должна, так как сторию прослушивания на десктопе я не использую. Использую только SpotifyHistoryPhone...
Или я что-то не понимаю?

**Reply by amd64fox** [reply]
У вас ошибка парсинга файла, скорее всего файл изменился и на андроиде, можете скинуть начала вашего файла `SpotifyHistoryPhone` сюда, скажу точнее тогда.

**Reply by Bondian** [reply]
Начало файла:
```
1655901600346#kbЋТ
   
video.subtitles""
player.license"on-demand"'
player.filter_explicit_contentfalse-
$player.filter_age_restricted_contentfalse
video.subtitles_ccfalse
audio.episode.speed1B
player.banned.artists)["spotify:artist:54QOSY1OciIWCfQFS0vLJ2"]Лэ
```
На всякий случай прикреплю файл, а то в превью комментария кодировка меняется на нечитаемые символы...
[SpotifyHistoryPhone.zip](https://github.com/Chimildic/goofy/files/8958696/SpotifyHistoryPhone.zip)

**Reply by amd64fox** [reply]
Да, файл изменился на новый, в нем нет истории прослушивания, поэтому и ошибка.

На андроиде пока не искал способ как вернуть старый вариант файла, но сразу же ясно, что можно выполнить откат на ту версию приложения с которой у вас работал скрипт и сидеть на ней пока не будет найден другой способ.

**Reply by Bondian** [reply]
Спасибо! Так и сделал. Откатился на старую весию (последующие версии не дают историю).
Может ещё кому пригодится: [Обновление от Google Play Market, Версия  8.7.20.1261 Build: 85467771 от 05.04.2022 на форуме 4pda](https://4pda.to/forum/index.php?showtopic=248440&st=53920#entry114237318)
Прикрепить здесь не получилось из-за ограничений размера файла...

**Comment by amd64fox** [comment]
Добавил предупреждение в шапке, пока в новых версиях я не нашел простого варианта брать историю, только откат на старые версии. 
Есть правда один сложный вариант это брать историю из файлов кеша, но там слишком сложно парсить файлы, если ничего не изменится то мб сделаю позже.
